Add support for flavor reservation

Mark-Powers opened the pull request:
Calendar screenshots:
<img width="889" alt="image" src="https://github.com/user-attachments/assets/17d4dbf1-8d2b-49ec-bab1-f3628317791f" />

<img width="889" alt="image" src="https://github.com/user-attachments/assets/c8aea7ea-3d2e-4171-9759-34e1855f0aa4" />

Mark-Powers committed:
Add support for flavor reservation

Mark-Powers committed:
Filter reservation table and fix switch class

Mark-Powers committed:
Improve gpu information on flavor res form

Mark-Powers commented on blazar_dashboard/static/leases/js/create_lease/extra_capability_widget.js:
```diff
@@ -16,6 +16,9 @@ function capabilitiesjs(resource_type, switchable_classname) {
     xhr.onreadystatechange = function() {
         if (this.readyState == 4 && this.status == 200) {
             capabilityValues = JSON.parse(this.responseText)['extra_capabilities'];
+            if (capabilityValues == undefined) {
```
This fixes a non-critical exception that would get logged sometimes. Not essential to this change.

Mark-Powers committed:
Fix minor issues

Mark-Powers committed:
Add availability calendar for flavors

msherman64 commented on blazar_dashboard/api/client.py:
```diff
@@ -491,6 +499,12 @@ def device_extra_capabilities(request):
 
+def flavors(request):
+    # NOTE we need >=v2.61 (Queens) for flavor description, which will be useful for users.
+    flavors = _nova.novaclient(request, version="2.61").flavors.list()
```
it feels wrong that we need to import a "private" nova method here.

msherman64 replied:
bleh, I actually agree we should do it this way, at least for now. Context:
[As per this bug](https://bugs.launchpad.net/horizon/+bug/1970619), horizon defaults to using the oldest compatible microversion, rather than the newest. :(

Since flavor descriptions were introduced in version 2.55, and "extra specs" to this api in 2.61,  calling either 2.55 or 2.61 should be fine here.

The "correct" way to handle this would be to [add a microversion "feature" to horizon](https://github.com/ChameleonCloud/horizon/blob/5c1523487a7cde65f658b2942d0a4b662b8dd9d0/openstack_dashboard/api/microversions.py#L29-L40), then to use the feature flag here, but that introduces even more coupling between horizon and our plugin, and might not get accepted upstream.

msherman64 reviewed (approved):
No obvious issues to me, just horizon being unintuitive w.r.t. microversions..

Mark-Powers committed:
Only show reservable flavors

Mark-Powers committed:
Fix trait string

Mark-Powers committed:
Revert "Fix trait string"

This reverts commit cf07403648ca76bfd5393585b6121bbb3c962296.

Mark-Powers committed:
Revert "Only show reservable flavors"

This reverts commit d0f192debb6d08b85fc5648e98bee4f4a26d2b6b.

Mark-Powers committed:
Add trait support

In the calendar availability calculation, check for trait compatibility.

Note: this requires placement policy to permit listing traits and
resource providers.

Mark-Powers committed:
Remove reserved flavors in dropdown

Mark-Powers committed:
Add check for traits with reservations

Mark-Powers committed:
Fix error if node_type not set in calendar

Mark-Powers committed:
Merge pull request #49 from ChameleonCloud/calendar_traits

Add support for calendar traits

Mark-Powers committed:
Fix requirements

Mark-Powers committed:
Filter out unreservable hosts in flavor calendar

Mark-Powers committed:
Filter flavor calendar by authorized projects

Mark-Powers committed:
Support host reservations in flavor chart

Mark-Powers committed:
Update to support CHAMELEON_ENABLE_VMS setting
